pco/src/data_types/floats.rs

```diff
@@ -6,144 +6,98 @@ use crate::constants::Bitlen;
-) -> PcoResult<(Mode<F::L>, Vec<Vec<F::L>>)> {
+) -> PcoResult<ModeAndLatents<F::L>> {
   match (
     chunk_config.float_mult_spec,
     chunk_config.float_quant_spec,
   ) {
-    (FloatMultSpec::Provided(_), FloatQuantSpec::Provided(_)) => {
-      return Err(PcoError::invalid_argument(
-        "FloatMult and FloatQuant cannot be provided (thus required) simultaneously",
-      ));
-    }
+    (FloatMultSpec::Provided(_), FloatQuantSpec::Provided(_)) => Err(PcoError::invalid_argument(
+      "FloatMult and FloatQuant cannot be provided (thus required) simultaneously",
+    )),
     (FloatMultSpec::Provided(base_f64), _) => {
       let base = F::from_f64(base_f64);
       let mode = Mode::float_mult(base);
-      let latents = float_mult_utils::split_latents(nums, base, base.inv());
+      let float_mult_config = FloatMultConfig {
+        base,
+        inv_base: base.inv(),
+      };
+      let latents = float_mult_utils::split_latents(nums, float_mult_config);
       Ok((mode, latents))
     }
-    (FloatMultSpec::Disabled, FloatQuantSpec::Provided(k)) => Ok((
+    (_, FloatQuantSpec::Provided(k)) => Ok((
       Mode::FloatQuant(k),
       float_quant_utils::split_latents(nums, k),
     )),
-    (FloatMultSpec::Enabled, FloatQuantSpec::Enabled) => {
-      let bids = compute_all_bids_float(nums);
-      let winning_bid = choose_winning_bid(&bids);
-      match winning_bid {
-        Bid::Candidate {
-          mode: mode @ Mode::FloatMult(base_l),
-          ..
-        } => {
-          let base = F::from_latent_ordered(*base_l);
-          let latents = float_mult_utils::split_latents(nums, base, base.inv());
-          Ok((*mode, latents))
-        }
-        Bid::Candidate {
-          mode: mode @ Mode::FloatQuant(k),
-          ..
-        } => {
-          let latents = float_quant_utils::split_latents(nums, *k);
-          Ok((*mode, latents))
-        }
-        Bid::Candidate {
-          mode: mode @ Mode::Classic,
-          ..
-        } => Ok((*mode, split_latents_classic(nums))),
-        Bid::Candidate { .. } => Err(PcoError::internal(
-          "Bug: float mode selection logic returned an int mode",
-        )),
-        Bid::Forfeit => Err(PcoError::internal(
-          "Bug: no candidate mode found, not even Classic",
-        )),
-      }
-    }
-    (FloatMultSpec::Enabled, _) => match float_mult_utils::compute_bid(nums) {
-      Bid::Candidate {
-        mode: mode @ Mode::FloatMult(base_l),
-        ..
-      } => {
-        let base = F::from_latent_ordered(base_l);
-        let latents = float_mult_utils::split_latents(nums, base, base.inv());
-        Ok((mode, latents))
-      }
-      Bid::Candidate { .. } => Err(PcoError::internal(
-        "Bug: expected a FloatMult bid, got a different mode",
-      )),
-      Bid::Forfeit => Ok((Mode::Classic, split_latents_classic(nums))),
-    },
-    (_, FloatQuantSpec::Enabled) => match sampling::choose_sample(
+    _ => Ok(detect_mode_and_split_latents(
```

Comment:
Looks like the casework on `Spec`s is split between this function and `detect_mode_and_split_latents`.  Would it make sense to put them all in one place, e.g. by handling `Provided` in the bid creation step?

Reply:
Would we just use a +1.0 or +Inf bid? I kinda like this approach better; seems a tad roundabout to make bids when we know which one will win.

Reply:
Oh right, that's a good reason to have separate code places -- if there's no auction, then handle it right here, and if there is an auction, then create the bids.  Sounds good.